Refactor background bridge connection options

Gudahtt opened the pull request:
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

The options passed into the background bridge have been refactored to distinguish between the different connection types more effectively.

Previously most constructor options were connection-specific, which made this module difficult to read. Now all connection-specific information has been grouped together and documented.

A few unused parameters were dropped, but this should include zero functional changes.

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [x] Any added code is fully documented

Gudahtt committed:
Refactor background bridge connection options

The options passed into the background bridge have been refactored to
distinguish between the different connection types more effectively.

Previously most constructor options were connection-specific, which
made this module difficult to read. Now all connection-specific
information has been grouped together and documented.

A few unused parameters were dropped, but this should include zero
functional changes.

Gudahtt commented on app/core/WalletConnect.js:
```diff
@@ -298,14 +298,14 @@ class WalletConnect {
-      webview: null,
-      url: this.hostname,
-      isWalletConnect: true,
-      wcWalletConnector: this.walletConnector,
```
This parameter was not referenced anywhere, so it has been removed.

Gudahtt replied:
Fixed here: 60afbfc5dec4c4848b25f2b4407a55f62feb4871

Gudahtt commented on app/core/BackgroundBridge/BackgroundBridge.js:
```diff
@@ -64,11 +90,22 @@ export class BackgroundBridge extends EventEmitter {
+    let portStream;
+    if (connectionOptions.type === CONNECTION_TYPE.WALLET_CONNECT) {
+      this.port = new WalletConnectPort(connectionOptions.wcRequestActions);
+      portStream = new MobilePortStream(this.port);
```
The port stream class for these first two connection types did not use the URL.

Gudahtt commented on app/core/BackgroundBridge/BackgroundBridge.js:
```diff
@@ -209,8 +247,13 @@ export class BackgroundBridge extends EventEmitter {
 
   notifySelectedAddressChanged(selectedAddress) {
-    if (this.isRemoteConn) {
-      if (!this.getApprovedHosts?.()?.[this.remoteConnHost]) return;
```
Note that I removed the optional chain operators here. The `SDKConnect.ts` module ensures that the approved host object is always defined and set to an object

Gudahtt commented on app/core/BackgroundBridge/BackgroundBridge.js:
```diff
@@ -27,31 +27,57 @@ const providerAsMiddleware = require('eth-json-rpc-middleware/providerAsMiddlewa
-    this.isMainFrame = isMainFrame;
-    this.isWalletConnect = isWalletConnect;
-    this.isRemoteConn = isRemoteConn;
-    this._webviewRef = webview && webview.current;
```
This condition was no longer needed (see below on line 192), because a ref is guaranteed to be passed in by the `BrowserTab` module.

andrepimenta reviewed (dismissed):
LGTM! 👌

christopherferreira9 commented:
Hi @Gudahtt 
The in app browser is not working on both platforms. Can you please check?

Gudahtt commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -930,8 +931,12 @@ export const BrowserTab = (props) => {
     const newBridge = new BackgroundBridge({
-      webview: webviewRef,
-      url: urlBridge,
+      connection: {
```
Oops, found a typo!
```suggestion
      connectionOptions: {
```

Gudahtt committed:
Fix typo

Gudahtt commented:
Thanks @christopherferreira9 great catch! I noticed it's because I made a typo for one of the constructor parameters. I have fixed that issue and checked the others for the same problem. Hopefully it is fixed now

christopherferreira9 commented:
I'm still seeing the same error. Here's what I can see in Metro with iOS:
`0 [TypeError: undefined is not an object (evaluating '_$$_REQUIRE(_dependencyMap[40], "../../../constants/urls").CONNECTION_TYPE.BROWSER')]`

If I manually change dApp in the in app browser tap I open the page but I can't interact with it.

Gudahtt commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -56,6 +56,7 @@ import { useTheme } from '../../../util/theme';
 import downloadFile from '../../../util/browser/downloadFile';
 import { createBrowserUrlModalNavDetails } from '../BrowserUrlModal/BrowserUrlModal';
 import {
+  CONNECTION_TYPE,
```
Oh my, this import is in the wrong place :facepalm: This is from the app constants not the URL constants. Will fix in a moment.

Gudahtt replied:
Fixed here: 60afbfc5dec4c4848b25f2b4407a55f62feb4871

Gudahtt committed:
Fix connection type import

Gudahtt commented:
Thanks again, time for round 3! :crossed_fingers:

Gudahtt commented:
Some issues remain - I will revisit this next week to debug further

BelfordZ commented on app/core/WalletConnect.js:
```diff
@@ -298,14 +298,14 @@ class WalletConnect {
-        rejectRequest: this.rejectRequest,
-        updateSession: this.updateSession,
+      connectionOptions: {
+        type: AppConstants.CONNECTION_TYPE.WALLET_CONNECT,
```
where do we use `CONNECTION_TYPE.BROWSER`? Is webview now unused?

Gudahtt replied:
It's used in the first file of this diff, `app/components/Views/BrowserTab`

Gudahtt commented:
Aha, I found the problem: some of the properties I removed here from `BackgroundBridge` were being used externally. I was only looking at internal usage. This should be easy to fix up; I'll restore any externally-used properties.

Gudahtt committed:
Restore `url` and `isMainFrame` properties

The `url` and `isMainFrame` properties were removed because they were
no longer used internally, but they are still used by the `BrowserTab`
component. They have been restored.

Gudahtt committed:
Fix JSDoc comment

Gudahtt commented:
I restored the `needs-dev-review` label so that someone can take a look at this commit specifically: https://github.com/MetaMask/metamask-mobile/pull/5616/commits/26466d721b55401a79db65faf9264911878ef19e
Which is an important change since the last dev review.

Gudahtt commented:
Seeing some e2e test errors here on Bitrise: https://app.bitrise.io/build/e6b6eb2a-ba00-4225-baf0-d9ae88bb6c20
Investigating

Gudahtt commented:
Removing `needs-dev-review` label until I have time to rebase this again
